### ISSUE TITLE
[FEATURE] RMCP+ support

### ISSUE THREAD

**Issue by starbops** [starbops/kubevirtbmc#30]
**Is your feature request related to a problem? Please describe.**

Currently, for the IPMI-over-LAN functionality, KubeVirtBMC only supports RMCP. But nowadays the mainstream is RMCP+, which is with the IPMIv2.

```
# IPMI 2.0
$ ipmitool -I lanplus -U admin -P password -H  172.19.31.70 power status
Error: Unable to establish IPMI v2 / RMCP+ session

# IPMI 1.5
$ ipmitool -I lan -U admin -P password -H  172.19.31.70 power status
Chassis Power is on
```

For example, [bmclib](https://github.com/bmc-toolbox/bmclib) does not support IPMI v1.5 at all. The flag `-I lanplus` is hardcoded in every `ipmitool` commands executed: https://github.com/bmc-toolbox/bmclib/blob/8a5f5a1a243c8f6174bb139a45c6cba30dabff9c/internal/ipmi/ipmi.go#L77

On the other hand, Tinkerbell Rufio leverages bmclib to communicate with target machine's BMC. As a result, KubeVirtBMC does not play well with Rufio. You'll get something like the following when adding a new Machine object:

```
{"level":"info","controller":"machine","controllerGroup":"bmc.tinkerbell.org","controllerKind":"Machine","Machine":{"name":"machine-1","namespace":"tink-system"},"namespace":"tink-system","name":"machi
ne-1","reconcileID":"9e871bba-03b2-42f2-bdb9-99b36ef83783","host":"172.19.31.70","username":"admin","v":0,"logger":"controllers/Machine","error":"8 errors occurred:\n\t* provider: asrockrack: failed to
 login\n\t* provider: supermicro: 404: failed to login\n\t* provider: gofish: 404: 404 page not found\n\n\t* provider: dell: 404: 404 page not found\n\n\t* provider: openbmc: not an OpenBMC device\n\t*
 provider: ipmitool: Error: Unable to establish IPMI v2 / RMCP+ session: exit status 1\n\t* provider: IntelAMT: unable to perform digest auth with http://172.19.31.70:16992/wsman: Post \"http://172.19.
31.70:16992/wsman\": context deadline exceeded\n\t* no Opener implementations found\n\n","providersAttempted":["gofish","dell","openbmc","ipmitool","asrockrack","IntelAMT","supermicro"],"successfulProv
ider":[],"caller":"/workspace/controller/client.go:39","time":"2024-10-23T09:14:24Z","message":"Failed to open connection to BMC"}
{"level":"error","controller":"machine","controllerGroup":"bmc.tinkerbell.org","controllerKind":"Machine","Machine":{"name":"machine-1","namespace":"tink-system"},"namespace":"tink-system","name":"mach
ine-1","reconcileID":"9e871bba-03b2-42f2-bdb9-99b36ef83783","error":"failed to open connection to BMC: 8 errors occurred:\n\t* provider: asrockrack: failed to login\n\t* provider: supermicro: 404: fail
ed to login\n\t* provider: gofish: 404: 404 page not found\n\n\t* provider: dell: 404: 404 page not found\n\n\t* provider: openbmc: not an OpenBMC device\n\t* provider: ipmitool: Error: Unable to estab
lish IPMI v2 / RMCP+ session: exit status 1\n\t* provider: IntelAMT: unable to perform digest auth with http://172.19.31.70:16992/wsman: Post \"http://172.19.31.70:16992/wsman\": context deadline excee
ded\n\t* no Opener implementations found\n\n","logger":"controllers/Machine","host":"172.19.31.70","caller":"/workspace/controller/machine.go:121","time":"2024-10-23T09:14:24Z","message":"BMC connectio
n failed"}
```

**Describe the solution you'd like**

Support RMCP+ protocol.

**Describe alternatives you've considered**

We can propose support for IPMI 1.5 for those projects. However, we don't know whether they are willing to support it, a fairly old standard that has been proven insecure.

**Additional context**

N/A

**Comment by starbops** [comment]
[This project](https://github.com/kurokobo/virtualbmc-for-vsphere) seems to support RMCP+ and might be worth looking at.

**Comment by halfcrazy** [comment]
Proposal to use https://github.com/bougou/go-ipmi archive lanplus support.

**Comment by halfcrazy** [comment]
hold until https://github.com/bougou/go-ipmi/pull/64 gets merged.

**Comment by starbops** [comment]
Interesting. Does [bougou/go-ipmi](https://github.com/bougou/go-ipmi) support server side implementation? I remember I searched for a while when initiating this project back then, there were no server side libraries except [vmware/goipmi](https://github.com/vmware-archive/goipmi) (sad that they've been publicly archived for some time).

**Comment by halfcrazy** [comment]
The server-side implementation of go-ipmi is not ready for out-of-the-box use and requires a great deal of custom development. We have already completed the POC in our internal branch, with around 1,500 new lines of IPMI code added.
The go-ipmi community is also working on this [feature](https://github.com/bougou/go-ipmi/pull/64). I prefer to wait for the official community implementation before integrating it into KubeVirtBMC. This will minimize code changes, rather than submitting our POC code upstream.

**Comment by halfcrazy** [comment]
There is an issue we need to discuss. The background is that the server-side of go-ipmi already supports IPMI v2, yet it cannot support IPMI v2 and v1.5 simultaneously.
Would we be okay with only implementing IPMI v2? (IPMI v1.5 has very few users at present, and most users are already on v2.)
However, I’m concerned this approach will break backward compatibility.

https://github.com/bougou/go-ipmi/blob/7892114510df1070bdef92cd28675c07b01c7817/pkg/server/server.go#L175-L195

**Comment by anishbista60** [comment]
> hold until bougou/go-ipmi#64 gets merged.

Seems this one is merged at upstream.

**Comment by halfcrazy** [comment]
> > hold until [bougou/go-ipmi#64](https://github.com/bougou/go-ipmi/pull/64) gets merged.
> 
> Seems this one is merged at upstream.

Yes, and I’ve also fixed several minor defects related to this MR.

**Comment by anishbista60** [comment]
> There is an issue we need to discuss. The background is that the server-side of go-ipmi already supports IPMI v2, yet it cannot support IPMI v2 and v1.5 simultaneously.
> Would we be okay with only implementing IPMI v2? (IPMI v1.5 has very few users at present, and most users are already on v2.)
> However, I’m concerned this approach will break backward compatibility.
> 
> https://github.com/bougou/go-ipmi/blob/7892114510df1070bdef92cd28675c07b01c7817/pkg/server/server.go#L175-L195

I'm not sure whether we have user heavily relying on IPMI v1.5. Most of the folks rely on redfish. Since, v2 comes with sever side auth + others enhancement. 

If user are negligible relying on v1.5, dropping it and adopting v2 would be way to go. 

But I would like to see other opinion as well. 

What's your opinion @halfcrazy @starbops ?

**Comment by halfcrazy** [comment]
My use case only involves IPMI v2, so dropping support for IPMI v1.5 would have no impact on me. Getting the go-ipmi community to maintain simultaneous support for both v1.5 and v2 would likely require a substantial amount of work.

**Comment by starbops** [comment]
I have no objection dropping IPMI v1.5 support. In fact, I was hoping that I could've done it with v2 at the inception of this project 😆

Let's do it. Just make sure we have the document updated and emphasized. Thank you.

**Comment by starbops** [comment]
Still, this will be a breaking change to the IPMI service our agent provides, so we should call it out on the release note for v0.9.0. I'll handle that part when doing the release.
